### PR TITLE
Refactored vm monitoring code

### PR DESCRIPTION
**Pull request by yuvalturg**
Moved common code to functions to ease the addition of new vmi metrics.

Signed-off-by: Yuval Turgeman <yturgema@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Review by maiqueb — COMMENTED** [review]
Some questions and comments.

Looks very good, nice job removing all this duplicated code.

File: pkg/monitoring/vms/prometheus/prometheus.go [hunk 1]
```diff
@@ -77,273 +77,127 @@ func tryToPushMetric(desc *prometheus.Desc, mv prometheus.Metric, err error, ch
...
-	for vcpuId, vcpu := range vmStats.Vcpu {
-		// Initial vcpu metrics labels
-		if !vcpu.StateSet || !vcpu.TimeSet {
-			log.Log.V(4).Warningf("State or time not set for vcpu#%d", vcpuId)
-		} else {
-			vcpuUsageLabels := []string{"node", "namespace", "name", "id", "state"}
-			vcpuUsageLabels = append(vcpuUsageLabels, metrics.k8sLabels...)
-			vcpuUsageDesc := prometheus.NewDesc(
+	for vcpuID, vcpu := range vmStats.Vcpu {
+		svcpuID := fmt.Sprintf("%v", vcpuID)
```
[the remark this answers is not in the document]
Comment (yuvalturg) [reply pkg/monitoring/vms/prometheus/prometheus.go]: Regarding the ID, IIUC, this is in fact the vcpuID, I fixed the other issues.

File: pkg/monitoring/vms/prometheus/prometheus.go [hunk 2]
```diff
@@ -356,106 +210,47 @@ func (metrics *vmiMetrics) updateNetwork(vmStats *stats.DomainStats) {
...
-				networkTrafficPktsRxLabelValues := []string{metrics.vmi.Status.NodeName, metrics.vmi.Namespace, metrics.vmi.Name, net.Name, "rx"}
-				networkTrafficPktsRxLabelValues = append(networkTrafficPktsRxLabelValues, metrics.k8sLabelValues...)
-
-				mv, err := prometheus.NewConstMetric(
-					networkTrafficPktsDesc, prometheus.CounterValue,
-					float64(net.RxPkts),
-					networkTrafficPktsRxLabelValues...,
-				)
-				tryToPushMetric(networkTrafficPktsDesc, mv, err, metrics.ch)
+				metrics.pushPrometheusMetric(desc, prometheus.CounterValue, float64(net.RxPkts), []string{net.Name, "rx"})
```
[the remark this answers is not in the document]
Comment (yuvalturg) [reply pkg/monitoring/vms/prometheus/prometheus.go:235]: Right, this refactoring is actually breaking #4646 to multiple PRs.  The first step would be this PR, and later on we will split some of these metrics to different names and not just different labels.

File: pkg/monitoring/vms/prometheus/prometheus.go [hunk 3]
```diff
@@ -677,12 +472,36 @@ type vmiMetrics struct {
 func (metrics *vmiMetrics) updateMetrics(vmStats *stats.DomainStats) {
 	metrics.updateKubernetesLabels()
 
-	metrics.updateMemory(vmStats)
+	metrics.updateMemory(vmStats.Memory)
```
[the remark this answers is not in the document]
Comment (yuvalturg) [reply pkg/monitoring/vms/prometheus/prometheus.go:475]: Fixed

File: pkg/monitoring/vms/prometheus/prometheus.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply pkg/monitoring/vms/prometheus/prometheus.go]: Hm, `vmStats.Vcpu` is an array: 
https://github.com/kubevirt/kubevirt/blob/ffe4040cb4205e3b4a9a19933586f44fefa507f5/pkg/virt-launcher/virtwrap/stats/types.go#L57

Of `DomainStatsVcpu`: 
https://github.com/kubevirt/kubevirt/blob/ffe4040cb4205e3b4a9a19933586f44fefa507f5/pkg/virt-launcher/virtwrap/stats/types.go#L72

AFAIU, when you iterate a slice(using the [range](https://golang.org/ref/spec#RangeClause) built-in), you get an index.

At the end of the day, this is not that important; the only question is if we *do* want to highlight if this parameter is an ID or not. I'm good either way. And I could also be wrong :)

[the remark this answers is not in the document]
Comment (yuvalturg) [reply pkg/monitoring/vms/prometheus/prometheus.go]: Yes, it's an array, I think that the index to the array serves as the vcpu id as well.  Either way, I changed it to Idx, here and in the block metrics as well

**Review by maiqueb — APPROVED** [review]
Thank you.

File: pkg/monitoring/vms/prometheus/prometheus.go [hunk 4]
```diff
@@ -77,385 +77,180 @@ func tryToPushMetric(desc *prometheus.Desc, mv prometheus.Metric, err error, ch
...
+			metrics.pushPrometheusMetric(desc, prometheus.GaugeValue, float64(mem.SwapOut)*1024, []string{"out"})
 		}
 	}
 }
 
-func (metrics *vmiMetrics) updateVcpu(vmStats *stats.DomainStats) {
-	for vcpuId, vcpu := range vmStats.Vcpu {
-		// Initial vcpu metrics labels
-		if !vcpu.StateSet || !vcpu.TimeSet {
-			log.Log.V(4).Warningf("State or time not set for vcpu#%d", vcpuId)
```
Comment (ashleyschuett) [inline pkg/monitoring/vms/prometheus/prometheus.go:187]: do we still want to have this warning log printed?

Comment (yuvalturg) [reply pkg/monitoring/vms/prometheus/prometheus.go:187]: Honestly, I don't think we need it @ashleyschuett - we don't print this warning in any other metric

**Pull request merged**